src/norm/mongodb.nim

```diff
@@ -0,0 +1,727 @@
+        proc_map[key] &= "      obj.$1 = none($2)\n".format(fieldName, center)
+      elif typeName.startsWith("seq["):
+        center = typeName.replace("seq[", "").replace("]", "")
+        proc_map[key] &= "  if doc.contains(\"$1\"):\n".format(bsonFieldName)
+        proc_map[key] &= "    obj.$1 = @[]\n".format(fieldName)
+        proc_map[key] &= "    for item in doc[\"$1\"].items:\n".format(bsonFieldName)
+        proc_map[key] &= "      if item.kind in @[$1]:\n".format(TYPE_TO_BSON_KIND[center])
+        proc_map[key] &= "        obj.$1.add item.$2\n".format(fieldName, TYPE_TO_BSON_PROC[center])
+      else:
+        discard
+    #
+    # now handle cross-object references
+    #
+    for field in obj.fields:
+      typeName = reconstructType(field.typ)
+      fieldName = field.signature.name
+      if NORM_UNIVERSAL_TYPE_LIST.contains(typeName):
+        continue
+      if normObjectNamesRegistry.contains(typeName):
+        proc_map[key] &= "  if doc.contains(\"$1\"):\n".format(fieldName)
+        proc_map[key] &= "    obj.$1 = $2()\n".format(fieldName, typeName)
+        proc_map[key] &= "    applyBSON(obj.$1, doc[\"$1\"])\n".format(fieldName)
+  #
+  # finish up all procedure strings
+  #
+  for key, s in proc_map.pairs():
+    result &= s
+    result &= "\n" # add a blank line between each proc
+
+
+macro db*(connection, user, password, database: string, body: untyped): untyped =
+  ##[
+    DB models definition. Models are defined as regular Nim objects in regular ``type`` sections.
+
+    ``connection``, ``user``, ``password``, ``database`` are the same args accepted by a standard ``dbConn`` instance.
+
+    The macro generates ``withDb`` template that wraps all DB interations.
+  ]##
+
+  result = newStmtList()
+
+  var dbObjReprs: seq[ObjRepr]
+  var normObjectNamesRegistry: seq[string] = @[] # this is later injected into context
+  var newCollections: seq[string]
+
+  for node in body:
+    if node.kind == nnkTypeSection:
+      let typeSection = node.ensureIdFields()
+      result.add typeSection
+
+      for typeDef in typeSection:
+        # echo typeDef.treeRepr
+        dbObjReprs.add typeDef.toObjRepr()
+
+    else:
+      result.add node
+
+  updateObjectRegistry(dbObjReprs)
+  for o in dbObjReprs:
+    newCollections.add o.getCollectionName
+
+  # echo $dbObjReprs
+  # echo $genObjectAccess(dbObjReprs)
+  # echo $genObjectToBSON(dbObjReprs)
+  # echo $genBSONToObject(dbObjReprs)
+
+
+  # let objectAccess =  parseStmt(genObjectAccess(dbObjReprs))
+  # result.add(objectAccess)
+
+  let withDbNode = getAst genWithDb(connection, user, password, database, -1, newCollections)  # -1 = true
+  result.insert(0, withDbNode)
+
+  let bsonToObject = parseStmt(genBSONToObject(dbObjReprs))
+  result.add(bsonToObject)
+
+  let objectToBSON = parseStmt(genObjectToBSON(dbObjReprs))
+  result.add(objectToBSON)
+
+macro dbAddCollection*(obj: typed): untyped =

```

Comment:
I like this idea, I can see how it can be useful. Probably, should be backported to other backends.

Reply:
I can help with that.

There is one gotcha to it that I will need both document and write a new Exception for: the Object must *already* have an 'id' field. Since the type already exists, it is not possible to add it dynamically. This needs to be caught at compile-time, so I will add a compile-time check.